bionic/code_references.py

```diff
@@ -0,0 +1,214 @@
+    # from which object an attribute is requested.
+    # Read more about bytecode at https://docs.python.org/3/library/dis.html
+
+    for op in dis.get_instructions(code):
+        try:
+            # Sometimes starts_line is None, in which case let's just remember the
+            # previous start_line (if any). This way when there's an exception we at
+            # least can point users somewhat near the line where the error stems from.
+            if op.starts_line is not None:
+                lineno = op.starts_line
+
+            if op.opname in ["LOAD_GLOBAL", "LOAD_NAME"]:
+                if op.argval in context.globals:
+                    set_tos(context.globals[op.argval])
+                else:
+                    # This can happen if the variable does not exist. If we
+                    # cannot find the variable, we return it's name instead.
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_DEREF", "LOAD_CLOSURE"]:
+                set_tos(context.cells[op.argval])
+            elif op.opname == "IMPORT_NAME":
+                # This instruction only appears if the code object imports a
+                # module using the import statement. If a user is importing
+                # modules inside a function, they probably don't want to import
+                # the module until the function execution time.
+                message = oneline(
+                    f"""
+                Bionic found a code reference in file ${code.co_filename} at line
+                ${lineno} that imports ${op.argval} module when hashing
+                ${code.co_name}. Changes inside the module will likely not be
+                detected. To detect changes inside the module, import it globally
+                instead.
+                """
+                )
+                warnings.warn(message)
+                set_tos(None)
+            elif op.opname in ["LOAD_METHOD", "LOAD_ATTR"]:
+                if isinstance(tos, str):
+                    tos += "." + op.argval
+                # TODO: Consider calling getattr only when TOS is a module.
+                # Doing so has risk of missing cases, and any missing case would
+                # be hard to detect for users. Before making this change,
+                # document all the cases that we won't be catching so that we are
+                # aware of them.
+                # TODO: Due to the same reason as why we don't import modules, we
+                # should not call `getattr` on properties. We should either add a
+                # separate check for properties, or implement the above TODO to
+                # fix this.
+                elif hasattr(tos, op.argval):
+                    tos = getattr(tos, op.argval)
+                else:
+                    refs.append(op.argval)
+            elif op.opname == "DELETE_FAST" and tos:
+                del context.varnames[op.argval]
```

Comment:
I hadn't noticed before that we're mutating the context object here. Should we make this clear in the docstring? Would we ever continue using the context object after calling this function?

Reply:
We discussed this sometime back (before I started working on the integration logic). I think we decided to not delete variables, so I'm going to remove `DELETE_FAST` branch from here.

Reply:
I think we should `DELETE_FAST` anyway, but just to be sure that there is no side effect of mutating the context object during this analysis phase, I'll just make a copy of `code_context` at the beginning of this method.